solr/solr-ref-guide/modules/deployment-guide/pages/collection-management.adoc

```diff
@@ -250,6 +250,7 @@ If `true` the states of individual replicas will be maintained as individual chi
 +
 Set core property _name_ to _value_.
 See the section xref:configuration-guide:core-discovery.adoc[] for details on supported properties and values.
```

Comment:
To confirm I understood well, are you mentioning the fact that we can add arbitrary properties (e.g., "foo" or "bar") this way?

Reply:
Your tests show arbitrary key-value.  Albeit your test doesn't show using them.   Hmmm.  I was thinking, maybe set `solr.directoryFactory` to `solr.MMapDirectoryFactory` (instead of the default NRT one) and then somehow detect it worked.  But this seems hard as Solr doesn't return it (AFAICT) although it prints some derivative info in the searcher stats that kind of reveals the choice.  Maybe if you were to add a static inner class superclass to another implementation (like solr.MMapDirectoryFactory) as pass-through, you could then check via some static boolean or something that it was instantiated?  Should be really short & simple.  For the value of the property, specify the full class name; can't use "solr." prefix/macro.

Reply:
i haven't had a chance to review the code, but this comment regarding testing (that arbitrary key=val options can be set values and actually impact the code) caught my eye.

the most straight forward way i can think of to test this would be with a custom config set that refrences those custom values in a way that's trivial to inspect (either from the CoreContainer or via request handler defaults that you could ask each replica for via an `echoParams=all&distrib=false` request)

A quick grep suggests we already have a `cloud-minimal-userproperties` configset with  `my.custom.prop` added by SOLR-16678 that' seems to exist exactly for this sort of purpose ... the tests added by that jira just don't check the 'persistence" aspect that this PR aims to fix.

Reply:
If I understood well, those arbitrary properties can later be used as substitutions. Another use (which were my original motivation) is that since they would now be stored in state.json, they can be later examined by plugins to inject custom behaviour. And pushing them inside the `core.properties` means they are available really easily without having to query the upper layers (replica/shard/collection) in a consistent way between cloud and non cloud setups.

Reply:
I also thought that having that inheritance mechanism (core inheriting properties from replica and collection) would make sense and make things more consistent. I found it awkward that `property.*` were pushed to the core and then completely discarded, meaning that the next replica added to the collection would not have those properties, unless explicitly specified in the ADDREPLICA command.